teos/src/api/internal.rs

```diff
@@ -280,27 +279,40 @@ impl PrivateTowerServices for Arc<InternalAPI> {
                 .map_or("an unknown address".to_owned(), |a| a.to_string())
         );
 
-        let mut matching_appointments = vec![];
-        let locator = Locator::from_slice(&request.into_inner().locator).map_err(|_| {
+        let req_data = request.into_inner();
+        let locator = Locator::from_slice(&req_data.locator).map_err(|_| {
             Status::new(
                 Code::InvalidArgument,
                 "The provided locator does not match the expected format (16-byte hexadecimal string)",
             )
         })?;
 
-        for (_, appointment) in self
+        let mut appointments: Vec<(UserId, Appointment)> = self
             .watcher
             .get_watcher_appointments_with_locator(locator)
+            .into_values()
+            .map(|appointment| (appointment.user_id, appointment.inner))
+            .collect();
+
+        let user_id_option = req_data.user_id_option;
+        if let Some(user_id_slice) = user_id_option {
+            let user_id = UserId::from_slice(&user_id_slice).map_err(|_| {
+                Status::new(
+                    Code::InvalidArgument,
+                    "The Provided user_id does not match expected format (33-byte hex string)",
+                )
+            })?;
+            appointments.retain(|(appointment_user_id, _)| *appointment_user_id == user_id);
```

Comment:
You are only filtering appointments in the watcher, but not trackers in the responder.

I better approach for this would be to pass the optional user_id to the db queries so we don't have to load unnecessary data. You have to do so both for `get_watcher_appointments_with_locator` and `get_responder_trackers_with_locator`